raiden-ts/src/transport/epics.ts

```diff
@@ -650,88 +696,98 @@ export const matrixPresenceUpdateEpic = (
+        )
+          // even if signature verification passes, this wouldn't change presence, so return early
+          return EMPTY;
 
-      // fetch profile info if user have no valid displayName set
-      const displayName$: Observable<string | undefined> = Signature.is(user.displayName)
-        ? of(user.displayName)
-        : defer(() => matrix.getProfileInfo(userId)).pipe(
-            pluck('displayname'),
-            catchError(() => of(undefined)),
-          );
+        // fetch profile info if user have no valid displayName set
+        const profile$ = Signature.is(user.displayName)
+          ? of({ displayname: user.displayName, avatar_url: user.avatarUrl })
+          : defer(() => matrix.getProfileInfo(userId));
 
-      return displayName$.pipe(
-        map(displayName => {
-          // errors raised here will be logged and ignored on catchError below
-          if (!displayName) throw new RaidenError(ErrorCodes.TRNS_NO_DISPLAYNAME, { userId });
-          // ecrecover address, validating displayName is the signature of the userId
-          const recovered = verifyMessage(userId, displayName) as Address | undefined;
-          if (!recovered || recovered !== address)
-            throw new RaidenError(ErrorCodes.TRNS_USERNAME_VERIFICATION_FAILED, {
-              userId,
-              receivedSignature: recovered!,
-            });
-          return recovered;
-        }),
-        // TODO: edge case: don't emit unavailable if address is available somewhere else
-        map(address =>
-          matrixPresence.success(
-            { userId, available, ts: user.lastPresenceTs ?? Date.now() },
-            { address },
-          ),
-        ),
-        catchError(err => (log.debug('Error validating presence event, ignoring', err), EMPTY)),
-      );
-    }),
-  );
+        return profile$.pipe(
+          map(profile => {
+            // errors raised here will be logged and ignored on catchError below
+            assert(profile?.displayname, 'no displayname');
+            // ecrecover address, validating displayName is the signature of the userId
+            const recovered = verifyMessage(userId, profile.displayname) as Address | undefined;
+            assert(
+              recovered === address,
+              `invalid displayname signature: ${recovered} !== ${address}`,
+            );
+            return matrixPresence.success(
+              {
+                userId,
+                available,
+                ts: user.lastPresenceTs ?? Date.now(),
+                caps: parseCaps(profile.avatar_url),
+              },
+              { address: recovered },
+            );
+          }),
+          catchError(err => (log.debug('Error validating presence event, ignoring', err), EMPTY)),
+        );
+      }),
+    )
+    .pipe(
+      withLatestFrom(latest$),
+      // filter out if presence update is to offline, and address became online in another user
```

Comment:
Nice edge case, I would have likely forgotten about it =)

Reply:
Thanks. Yes, shouldn't be too serious on our test environment, as we have a single server, but on  federated environments, quickly switching servers could give users unexpected bugs of appearing offline.